Test file locking

nickvergessen opened the pull request:
- [ ] Depends on: Testing app fake locks: https://github.com/owncloud/core/pull/22823
- [x] Test skips for broken client versions (https://github.com/owncloud/smashbox/pull/134#issuecomment-196832005)
- [x] Test skips for non DB-locking backends
### Test description

| Step | User |
| --- | --- |
| 2 | Enable QA testing app |
| 3 | Create dir/subdir/ |
| 4 | Populate locks |
| 5 | Try to upload dir/subdir/file2.dat |
| 6 | Remove locks |
| 7 | Upload dir/subdir/file2.dat |
### Cases

| Case | Description | Result | Test |
| --- | --- | --- | --- |
| 0 | Exclusive lock on dir/ | Can not upload | :x: - File upload fails in 7 as well |
| 1 | Shared lock on dir/ | Can upload | :white_check_mark: |
| 2 | Exclusive lock on dir/subdir/ | Can not upload | :x: - File upload fails in 7 as well |
| 3 | Shared lock on dir/subdir/ | Can upload | :white_check_mark: |
| 4 | Exclusive lock on dir/ + Shared lock on dir/subdir/ | Can not upload | :x: - File upload fails in 7 as well |
| 5 | Shared lock on dir/ + Exclusive lock on dir/subdir/ | Can not upload | :x: - File upload fails in 7 as well |
| 6 | Shared lock on dir/ and dir/subdir/ | Can upload | :white_check_mark: |

nickvergessen committed:
Add a test for the new locking backend

nickvergessen commented on lib/owncloud/test_locking.py:
```diff
@@ -0,0 +1,113 @@
+
+    for lock in use_locks:
+        try:
+            oc_api.make_ocs_request(
```
@PVince81 I added this method to the pyoc client:
```py
    def make_ocs_request(self, method, service, action, accepted_codes=[100]):
        res = self.__make_ocs_request(method, service, action)
        if res.status_code == 200:
            tree = ET.fromstring(res.content)
            self.__check_ocs_status(tree, accepted_codes=accepted_codes)
            return res
        raise OCSResponseError(res)
```

I think it makes sense to have something like this in place. Otherwise I have to do this on my own in here.
What do you think?

PVince81 replied:
Yeah, a generic OCS method makes sense

nickvergessen replied:
https://github.com/owncloud/pyocclient/pull/126

nickvergessen commented on lib/owncloud/test_locking.py:
```diff
@@ -0,0 +1,136 @@
+    lock_provider.unlock()
+
+
+#  TODO move the block below into cernbox
```
https://github.com/cernbox/smashbox/issues/56

nickvergessen commented:
This PR is basically done now, open questions:
* Should we move https://github.com/nickv-oc/files_lockprovisioning to the owncloud org? @DeepDiver1975 
* Is it okay if this only works with database testing? @DeepDiver1975

nickvergessen commented:
Okay, I seem to have found out the issue.
When we sync the folders in step 3, they don't have a md5 (etag)set, also the folders have `RDNVW` as permission.
when we now do a second sync, the permissions are set correctly to `RDNVCK`.
But when we already moved a file into the folder, the file will fail because `C` is not in the permissions.

Not sure if "adding a second sync" is the way to go, or if the client should update the etag and permissions on the first sync?

guruz commented:
> When we sync the folders in step 3, they don't have a md5 (etag)set,

That's normal and intended, the next sync would update them via PROPFIND.

> also the folders have RDNVW as permission.

That permission string is created by who? The server? Because I don't see it retrieved in `PropagateRemoteMkdir` and I don't find in the client where that could be created

FYI @ogoffart @ckamm

nickvergessen commented:
> > also the folders have RDNVW as permission.
> 
> That permission string is created by who? The server? 

I don't think so, because a propget always says permissions `RDNVCK` for me. So I assume that is client discovery.

nickvergessen commented:
Here are the full sync logs of a try:
https://gist.github.com/nickvergessen/8f5b8ea746a32a7a934b

nickvergessen commented:
Okay, so I changed the test to work with folders only, instead of files. Tests still failed, then I removed the skeleton files, to make sure, that the test always only ever deals with files, and see there: test passes.

So it seems that some variable is not reset correctly, after dealing with a file before reusing it for a folder. (All cheers to @guruz for that hint)

nickvergessen commented:
I can't seem to find the root cause of this. I can only confirm that two things fix the issue:
1. Don't block the first sync of the file
2. Don't sync a files but folders only

I hope that helps the desktop guy stepping in here.

ckamm commented:
@nickvergessen  A potential fix was merged to master 7bd4f95b8c200e0fc2d5eff2bc5ab2428356a2ba, see owncloud/client#4548

nickvergessen commented:
Yes, confirmed that it works with that commit. Nice find!

nickvergessen committed:
Detect the fixed client version and locking

nickvergessen commented:
The test is written in a way, that it will not fail on 9.0 and when using another locking then DB, so it should be good to merge @PVince81

PVince81 commented:
:+1:

nickvergessen committed:
Move to owncloudorg folder
